chirpstack/src/storage/sqlite.rs

```diff
@@ -0,0 +1,180 @@
+use std::sync::RwLock;
+use std::time::Instant;
+
+use anyhow::Result;
+use tracing::info;
+
+use crate::monitoring::prometheus;
+use diesel::sqlite::SqliteConnection;
+use diesel::{Connection, ConnectionError, ConnectionResult};
+use diesel_async::pooled_connection::deadpool::{Object as DeadpoolObject, Pool as DeadpoolPool};
+use diesel_async::pooled_connection::{AsyncDieselConnectionManager, ManagerConfig};
+use diesel_async::sync_connection_wrapper::SyncConnectionWrapper;
+use futures::future::{BoxFuture, FutureExt, TryFutureExt};
+use scoped_futures::ScopedBoxFuture;
+use prometheus_client::metrics::histogram::{exponential_buckets, Histogram};
+
+use crate::config;
+
+pub type AsyncSqlitePool = DeadpoolPool<SyncConnectionWrapper<SqliteConnection>>;
+pub type AsyncSqlitePoolConnection = DeadpoolObject<SyncConnectionWrapper<SqliteConnection>>;
+
+lazy_static! {
+    static ref ASYNC_SQLITE_POOL: RwLock<Option<AsyncSqlitePool>> = RwLock::new(None);
+    static ref STORAGE_SQLITE_CONN_GET: Histogram = {
+        let histogram = Histogram::new(exponential_buckets(0.001, 2.0, 12));
+        prometheus::register(
+            "storage_sqlite_conn_get_duration_seconds",
+            "Time between requesting a SQLite connection and the connection-pool returning it",
+            histogram.clone(),
+        );
+        histogram
+    };
+}
+
+pub fn setup(conf: &config::Sqlite) -> Result<()> {
+    info!("Setting up SQLite connection pool");
+    let mut config = ManagerConfig::default();
+    config.custom_setup = Box::new(sqlite_establish_connection);
+    let mgr =
+        AsyncDieselConnectionManager::<SyncConnectionWrapper<SqliteConnection>>::new_with_config(
+            &conf.path, config,
+        );
+    let pool = DeadpoolPool::builder(mgr)
+        .max_size(conf.max_open_connections as usize)
+        .build()?;
+    set_async_db_pool(pool);
+
+    Ok(())
+}
+
+fn sqlite_establish_connection(
+    url: &str,
+) -> BoxFuture<ConnectionResult<SyncConnectionWrapper<SqliteConnection>>> {
+    let url = url.to_string();
+    tokio::task::spawn_blocking(
+        move || -> ConnectionResult<SyncConnectionWrapper<SqliteConnection>> {
+            let mut conn = SqliteConnection::establish(&url)?;
+
+            use diesel::connection::SimpleConnection;
+            // A few settings need to be configure:
+            // - enable foreign keys since it's off by default in sqlite
+            // - set busy_timeout to avoid manually managing transaction business/contention
+            // see https://sqlite.org/rescode.html#busy
+            // - set user journal mode
+            let conf = config::get();
+            let pragmas = pragma_generate(&conf.sqlite);
```

Comment:
Instead of using `pragma_generate`, we could get the list of pragmas from the config. E.g. we would have in the config:

```
pragmas = [
    "journal_mode = XXX",
    "busy_timeout = YYYY",
    "foreign_keys = ON",
    ...
]
```

Would that make sense instead of specifying each config option within ChirpStack? E.g. if some options are missing (e.g. foreign-keys) ChirpStack might not work correctly, but on the other hand, some other pragmas can't be configured currently. We also want to avoid that each pragma turns into a ChirpStack option.

Reply:
We would provide sane defaults and if users want to override, they would have the option to do so.